lib/SILOptimizer/Differentiation/LinearMapInfo.cpp

```diff
@@ -331,10 +333,24 @@ void LinearMapInfo::generateDifferentiationDataStructures(
   }
 
   // Add linear map fields to the linear map tuples.
-  for (auto &origBB : *original) {
+  // Now we need to be very careful as we're having a very subtle
+  // chicken-and-egg problem. We need lowered branch trace enum type for the
```

Comment:
Some nits to make documentation clearer.

> However its type lowering depends on its elements (at very least, the type classification of being trivial / non-trivial). 
I believe "its" here is the linear map tuple? Can we specify that explicitly?

> As the lowering is cached
"lowering" here is the Linear map tuple lowering? Can we specify that explicitly?

> we cannot add more entries without causing some very subtle issues later on
More entries to the branch trace enum (as in more cases) or more entries to the linear map tuple?

> Traverse all BBs in reverse post-order traversal order to ensure we process each BB before its predecessors.
May be its worth calling out that RPOT is not some variant of post-order traversal, rather the literal reversed order nodes obtained from post-order traversal.

Reply:
> I believe "its" here is the linear map tuple? Can we specify that explicitly?

No, the opposite.

> "lowering" here is the Linear map tuple lowering? Can we specify that explicitly?

No, we are talking about branch trace enums here.

> More entries to the branch trace enum (as in more cases) or more entries to the linear map tuple?

See above.

> May be its worth calling out that RPOT is not some variant of post-order traversal, rather the literal reversed order nodes obtained from post-order traversal.

RPOT is quite a standard technique for data flow problems and is discussed in every compiler course.

Reply:
> "lowering" here is the Linear map tuple lowering? Can we specify that explicitly?
No, we are talking about branch trace enums here.

A quick question for my own curiosity. Where do we cache the lowered types of the branch trace enums?

Reply:
It's the part of TypeLowering.